cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerSyncApiCommandIdProvider.java

```diff
@@ -0,0 +1,95 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import com.cloudera.api.swagger.ClustersResourceApi;
+import com.cloudera.api.swagger.client.ApiException;
+import com.cloudera.api.swagger.model.ApiCommand;
+import com.sequenceiq.cloudbreak.cm.commands.SyncApiCommandPollerConfig;
+import com.sequenceiq.cloudbreak.cm.commands.SyncApiCommandRetriever;
+import com.sequenceiq.cloudbreak.cm.polling.ClouderaManagerPollingServiceProvider;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+
+@Service
+public class ClouderaManagerSyncApiCommandIdProvider {
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(ClouderaManagerSyncApiCommandIdProvider.class);
+
+    private final SyncApiCommandPollerConfig syncApiCommandPollerConfig;
+
+    private final SyncApiCommandRetriever syncApiCommandRetriever;
+
+    private final ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider;
+
+    private final ExecutorService executorService;
+
+    public ClouderaManagerSyncApiCommandIdProvider(
+            SyncApiCommandRetriever syncApiCommandRetriever,
+            ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider,
+            SyncApiCommandPollerConfig syncApiCommandPollerConfig,
+            @Qualifier("cloudbreakListeningScheduledExecutorService")
+                    ExecutorService executorService) {
+        this.syncApiCommandRetriever = syncApiCommandRetriever;
+        this.executorService = executorService;
+        this.clouderaManagerPollingServiceProvider = clouderaManagerPollingServiceProvider;
+        this.syncApiCommandPollerConfig = syncApiCommandPollerConfig;
+    }
+
+    /**
+     * Obtain a sync API client config command ID with the following steps:
+     * 1. check are there any existing successful sync API commands
+     * 2. execute the sync API command with a timeout
+     * 3. if the API command ties out, use 1. with polling
+     * 4. if polling finished, do a last check, returns with the command id or throw an exception
+     */
+    public BigDecimal executeSyncApiCommandAndGetCommandId(String commandName, ClustersResourceApi api, Stack stack,
+            Callable<ApiCommand> callable) throws CloudbreakException, ApiException {
+        Optional<BigDecimal> lastSyncApiCommand = syncApiCommandRetriever.getCommandId(commandName, api, stack, true);
+        if (lastSyncApiCommand.isPresent()) {
+            LOGGER.debug("Found already existing {} command with id: {}",
+                    commandName, lastSyncApiCommand);
+        }
+        Future<ApiCommand> future = executorService.submit(callable);
```

Comment:
this part is a bit odd for me. So if we found a running deployclientconfig we would still send the command to CM? I mean `syncApiCommandRetriever.getCommandId` won't check for running/active commands. In the old implementation we always start checking if there is an already running command

Reply:
right, actually that would cause problems if at some point one command become async from sync

Reply:
btw i have just added this for the only usage where we are checking the already running commands (in order to not change any existing behaviour), would make sense to do this for activate parcels + configure kerberos